airbyte-integrations/connectors/source-google-ads/source_google_ads/custom_query_stream.py

```diff
@@ -0,0 +1,158 @@
+
+        google_datatype_mapping = {
+            "INT64": "integer",
+            "INT32": "integer",
+            "DOUBLE": "number",
+            "STRING": "string",
+            "BOOLEAN": "boolean",
+            "DATE": "string",
+        }
+        fields = CustomQuery.get_query_fields(self.user_defined_query)
+        fields.append(self.cursor_field)
+        google_schema = self.google_ads_client.get_fields_metadata(fields)
+
+        for field in fields:
+            node = google_schema.get(field)
+            # Data type return in enum format: "GoogleAdsFieldDataType.<data_type>"
+            google_data_type = str(node.data_type).replace("GoogleAdsFieldDataType.", "")
+            if google_data_type == "ENUM":
+                field_value = {"type": "string", "enum": list(node.enum_values)}
+            elif google_data_type == "MESSAGE":  # this can be anything (or skip as additionalproperties) ?
+                output_type = ["string", "number", "array", "object", "boolean", "null"]
+                field_value = {"type": output_type}
+            else:
+                output_type = [google_datatype_mapping.get(google_data_type, "string"), "null"]
+                field_value = {"type": output_type}
+            local_json_schema["properties"][field] = field_value
+
+        return local_json_schema
+
+    # Regexp flags for parsing GAQL query
+    RE_FLAGS = re.DOTALL | re.MULTILINE | re.IGNORECASE
+    # Regexp for getting query columns
+    SELECT_EXPR = re.compile("select(.*)from", flags=RE_FLAGS)
+    WHERE_EXPR = re.compile("where.*", flags=RE_FLAGS)
+    # list of keywords that can come after WHERE clause,
+    # according to https://developers.google.com/google-ads/api/docs/query/grammar
+    KEYWORDS_EXPR = re.compile("(order by|limit|parameters)", flags=RE_FLAGS)
+
+    @staticmethod
+    def get_query_fields(query: str) -> List[str]:
+        fields = CustomQuery.SELECT_EXPR.search(query)
+        if not fields:
+            return []
+        fields = fields.group(1)
+        return [f.strip() for f in fields.split(",")]
+
+    @staticmethod
+    def insert_segments_date_expr(query: str, start_date: str, end_date: str) -> str:
+        """
+        Insert segments.date condition to break query into slices for incremental stream.
+        :param query Origin user defined query
+        :param start_date start date for metric (inclusive)
+        :param end_date end date for metric (inclusive)
+        :return Modified query with date window condition included
+        """
+        # insert segments.date field
+        columns = CustomQuery.SELECT_EXPR.search(query)
+        if not columns:
+            raise Exception("Not valid GAQL expression")
+        columns = columns.group(1)
+        new_columns = columns + ", segments.date\n"
+        result_query = query.replace(columns, new_columns)
+
+        # Modify/insert where condition
+        where_cond = CustomQuery.WHERE_EXPR.search(result_query)
```

Comment:
this is a very cool approach, I like it. The only issue I see here is `WHERE` could contain a time-bound function like `LAST_14_DAYS` etc. then wouldn't it conflict with this incremental logic? 

In this case I think we should not add the where condition and just default to what the user did

Reply:
In this case I've added checking if custom query contain "segments.date" field on check_connection method. It wont allow adding time-bound conditions to custom queries.